parcels/tools/id_generators.py

```diff
@@ -0,0 +1,1036 @@
+        recovering the ID, reusing it for the next 'getID(...)' call
+        :arg id: ID to be released (and possibly recovered)
+        """
+        full_bits = np.uint32(4294967295)
+        nil_bits = np.int32(0)
+        spatiotemporal_id = np.bitwise_and(np.bitwise_or(np.left_shift(np.int64(full_bits), 32), np.int64(nil_bits)), np.int64(id))
+        spatiotemporal_id = np.uint32(np.right_shift(spatiotemporal_id, 32))
+        local_id = np.bitwise_and(np.bitwise_or(np.left_shift(np.int64(nil_bits), 32), np.int64(full_bits)), np.int64(id))
+        local_id = np.uint32(local_id)
+        self._release_id(spatiotemporal_id, local_id)
+
+    def __len__(self):
+        """
+        :returns the number of generated IDs
+        """
+        return np.sum(self.local_ids) + sum([len(entity) for entity in self.released_ids])
+
+    def get_length(self):
+        """
+        function forward to
+        :returns the number of generated IDs
+        """
+        return self.__len__()
+
+    def _get_next_id(self, lon_index, lat_index, depth_index, time_index=None):
+        """
+        private function - generates a conclusive 64-bit ID from clamped lon-lat-depth-time indices
+        """
+        local_index = -1
+        lon_shift = 32-int(np.ceil(np.log2(self._lon_bins)))
+        lat_shift = lon_shift-int(np.ceil(np.log2(self._lat_bins)))
+        id = np.left_shift(lon_index, lon_shift) + np.left_shift(lat_index, lat_shift) + depth_index
+        if len(self.released_ids) > 0 and (id in self.released_ids.keys()) and len(self.released_ids[id]) > 0:
+            local_index = np.uint32(self.released_ids[id].pop())
+            if len(self.released_ids[id]) <= 0:
+                del self.released_ids[id]
+        else:
+            local_index = self.local_ids[lon_index, lat_index, depth_index]
+            self.local_ids[lon_index, lat_index, depth_index] += 1
+        id = np.int64(id)
+        id = np.bitwise_or(np.left_shift(id, 32), np.int64(local_index))
+        id = np.uint64(id)
+        if self._track_id_index:
+            self._map_id_totalindex[id] = self._total_ids
+        self._total_ids += 1
+        self._used_ids += 1
+        return id
+
+    def _release_id(self, spatiotemporal_id, local_id):
+        """
+        private function - releases a 64-bit ID by adding the 32-bit local ID to the spatial lon-lat-depth-time ID map
+        :arg spatiotemporal_id: 32-bit spatial ID segment
+        :arg local_id: 32-bit sequential ID segment
+        """
+        if not self._recover_ids:
+            return
+        if spatiotemporal_id not in self.released_ids.keys():
+            self.released_ids[spatiotemporal_id] = []
+        self.released_ids[spatiotemporal_id].append(local_id)
+        self._used_ids -= 1
+
+
+class SpatioTemporalIdGenerator(BaseIdGenerator):
+    """
+    Generates 64-bit IDs based on 4D spatiotemporal coordinates, aiming for an ID distribution so that spatially-close
+    items have also close-by IDs.
```

Comment:
What would be a usecase where particles also need to be close in time? Since Parcels uses one timestepping scheme for all particles (so that they always all live on the same timestep), what is the difference between this class and the spatial 3D class?

Reply:
Still needs a response?